Source/EasyNetQ/Internals/AsyncQueue.cs

```diff
@@ -0,0 +1,129 @@
+        ///     This is an internal API that supports the EasyNetQ infrastructure and not subject to
+        ///     the same compatibility as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new EasyNetQ release.
+        /// </summary>
+        public AsyncQueue()
+        {
+        }
+
+        /// <summary>
+        ///     This is an internal API that supports the EasyNetQ infrastructure and not subject to
+        ///     the same compatibility as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new EasyNetQ release.
+        /// </summary>
+        public AsyncQueue(IEnumerable<T> collection)
+        {
+            foreach (var element in collection)
+                elements.Enqueue(element);
+        }
+
+        /// <summary>
+        /// Returns count of elements in queue
+        /// </summary>
+        public int Count
+        {
+            get
+            {
+                lock (mutex)
+                    return elements.Count;
+            }
+        }
+
+        /// <summary>
+        /// Dequeue element from queue
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token</param>
+        /// <returns>The dequeued element</returns>
+        public Task<T> DequeueAsync(CancellationToken cancellationToken = default)
+        {
+            lock (mutex)
+            {
+                if (elements.Count > 0)
+                    return Task.FromResult(elements.Dequeue());
+
+                CleanUpCancelledWaiters();
+
+                var waiter = new TaskCompletionSource<T>(TaskCreationOptions.RunContinuationsAsynchronously);
+                waiter.AttachCancellation(cancellationToken);
+                waiters.Enqueue(waiter);
+                return waiter.Task;
+            }
+        }
+
+        /// <summary>
+        /// Enqueue element to queue
+        /// </summary>
+        /// <param name="element">The element to enqueue</param>
+        public void Enqueue(T element)
+        {
+            lock (mutex)
+            {
+                var wasAttachedToWaiter = false;
+                while (waiters.Count > 0)
+                {
+                    var waiter = waiters.Dequeue();
+                    if (waiter.TrySetResult(element))
+                    {
+                        wasAttachedToWaiter = true;
+                        break;
+                    }
+                }
+
+                CleanUpCancelledWaiters();
+
+                if (wasAttachedToWaiter)
+                    return;
+
+                elements.Enqueue(element);
+            }
+        }
```

Comment:
It seems to me that this variant of the method works in exactly the same way but without additional variable (i.e. simpler).

```suggestion
        public void Enqueue(T element)
        {
            lock (mutex)
            {
                while (waiters.Count > 0)
                {
                    var waiter = waiters.Dequeue();
                    if (waiter.TrySetResult(element))
                    {
                        CleanUpCancelledWaiters();
                        return;
                    }
                }

                elements.Enqueue(element);
            }
        }
```